pkg/controller/direct/compute/firewallpolicyrule/firewallpolicyrule_controller.go

```diff
@@ -0,0 +1,297 @@
+}
+
+var _ directbase.Adapter = &firewallPolicyRuleAdapter{}
+
+func (m *firewallPolicyRuleModel) client(ctx context.Context) (*gcp.FirewallPoliciesClient, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewFirewallPoliciesRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building FirewallPolicy client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *firewallPolicyRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ComputeFirewallPolicyRule{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Set label managed-by-cnrm: true
+	obj.ObjectMeta.Labels["managed-by-cnrm"] = "true"
+
+	// Get firewall policy
+	firewallPolicyRef, err := ResolveComputeFirewallPolicy(ctx, reader, obj, obj.Spec.FirewallPolicyRef)
+	if err != nil {
+		return nil, err
+
+	}
+	obj.Spec.FirewallPolicyRef.External = firewallPolicyRef.External
+	firewallPolicy := obj.Spec.FirewallPolicyRef.External
+
+	// Get priority
+	priority := obj.Spec.Priority
+
+	firewallPolicyRuleAdapter := &firewallPolicyRuleAdapter{
+		firewallPolicy: firewallPolicy,
+		priority:       priority,
+		desired:        obj,
+		reader:         reader,
+	}
+
+	// Get GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("building gcp client: %w", err)
+	}
+	firewallPolicyRuleAdapter.firewallPoliciesClient = gcpClient
+
+	return firewallPolicyRuleAdapter, nil
+}
+
+func (m *firewallPolicyRuleModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeFirewallPolicyRule", "priority", a.priority)
+
+	firewallPolicyRule, err := a.get(ctx)
+	if err != nil {
+		// When a certain rule does not exist, the error has code 400(bad request) instead of 404(not found)
+		// example error message:
+		// "Invalid value for field 'priority': '9000'. The firewall policy does not contain a rule at priority 9000.",
+		if direct.IsBadRequest(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting ComputeFirewallPolicyRule %d: %w", a.priority, err)
+	}
+	a.actual = firewallPolicyRule
+	return true, nil
+}
+
+func (a *firewallPolicyRuleAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	var err error
+
+	err = resolveDependencies(ctx, a.reader, a.desired)
+	if err != nil {
+		return err
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeFirewallPolicyRule", "priority", a.priority)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
```

Comment:
Do we need to DeepCopy?

Reply:
We use DeepCopy() to avoid unintended modifications to the original a.desired. However, I do not see we currently modify this object anywhere later in the code... The DeepCopy() might be unnecessary, but it's no harm to keep it(?) just in case... since we don't want to change any desired values...